### PR TITLE
feat(http): adds support for generic http requests

### PR DESCRIPTION
**Pull request by nathanielc**
This commits add a new package experimental/http/requests API to allow for generic
request response interactions over HTTP.

This does not change existing http.get or http.post as there are
thousands of live production queries using those methods.

The plan is to use this new requests package for a few weeks/months in
experimental and shake out all the bugs. Then promote it out of
experimental and then add deprecation warnings to the existing
functions.

Fixes: https://github.com/influxdata/flux/issues/3653

### Done checklist
- [x] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: stdlib/experimental/http/requests/requests.go [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+				// Alias the DNS lookup error so as not to disclose the
+				// DNS server address. This error is private in the net/http
+				// package, so string matching is used.
+				if strings.HasSuffix(err.Error(), "no such host") {
+					return 0, nil, nil, errors.New(codes.Invalid, "no such host")
+				}
+				return 0, nil, nil, err
+			}
+			body, err := ioutil.ReadAll(response.Body)
+			_ = response.Body.Close()
```
Comment (skartikey) [inline stdlib/experimental/http/requests/requests.go:172]: if `response.Body` is nil, the code will panic with the invalid memory address or nil pointer dereference. So I guess moving this code line after `nil` error check below will make sure that the code returns if there's any error returned from `ioutil.ReadAll` or we can put a nil check on `response.Body` before calling Close on it.

**Review by skartikey — CHANGES_REQUESTED** [review]
Looks good to me with 1 minor change mentioned below.

File: stdlib/experimental/http/requests/requests.go (hunk 1 above)
Comment (nathanielc) [reply stdlib/experimental/http/requests/requests.go:172]: @skartikey and I chatted out of band but Go guarantees the Body is non-nil, see https://pkg.go.dev/net/http#Response

**Review by skartikey — APPROVED** [review]

**Comment by nathanielc** [comment]
One design question I had is should we make params take a list of values instead of just a single value?

Specifically should the API for `do` become?

```
builtin _do : (
        method: string,
        url: string,
        // The PR current has ?params: [string: string]
        ?params: [string:[string]],
        ?headers: [string:string],
        ?body: bytes,
        config: {A with timeout: duration, insecureSkipVerify: bool},
    ) => {statusCode: int, body: bytes, headers: [string:string]}
```

**Review by onelson — COMMENTED** [review]
Just looking at the high level, the docs show much of what I'd expect to see.

I'm not sure what the default HTTP client behavior is, but somethings I might expect to configure:

- redirect policy (follow or not)
- loading a custom CA (which is more desirable than skipping cert verification)

There is some supporting API I'd likely reach for while using this which isn't included. I'm not sure if it belongs here or in a separate package.

Concretely:
- functions to help build urls (specifically to help URI encode path segments and query params)
- something to base64 encode strings (commonly needed to prepare headers/bodies)

I forget how python's `requests` handles encoding query params (if it's automatic or if the caller is expected to do it themselves). I'd weakly suspect it's the caller's job since the server implementation in question could be one that doesn't adhere to the HTTP spec (it happens), so callers may need ways to "do the wrong thing on purpose."

I think there is a separate ticket out there related to hashing (md5, sha) which is often needed for making REST API requests. Common need is to hash some parts of the request params _with the current timestamp_, so it has to be done as close to the request being sent as possible.

Food for thought.

**Comment by onelson** [comment]
> One design question I had is should we make params take a list of values instead of just a single value?

I must have overlooked this in my skimming of the diff! Being able to support multiple occurrences of a given key is important. Commonly I've seen params expressed as a sequence of tuple pairs. If a map of `string: [string]` makes more sense for flux, then I'd take it, but in other langs I'd expect to be able to push new pairs into an indexed collection.

File: stdlib/experimental/http/requests/requests.flux [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        params: params,
+        headers: headers,
+        body: body,
+        config: config,
+    )
+
+// get makes a http GET request. This identical to calling `request.do(method: "GET", ...)`.
+//
+// ## Parameters
+// - url: url to request, this should not include any query parameters.
```
Comment (noramullen1) [inline stdlib/experimental/http/requests/requests.flux:178]:
```suggestion
// - url: URL to request. This should not include any query parameters.
```

File: stdlib/experimental/http/requests/requests.flux [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+// - params: params is a set of key value pairs to add to the URL as query parameters.
```
Comment (noramullen1) [inline stdlib/experimental/http/requests/requests.flux:179]:
```suggestion
// - params: Set of key value pairs to add to the URL as query parameters.
```

File: stdlib/experimental/http/requests/requests.flux [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+// - headers: headers is a set of key values pairs to include on the request
```
Comment (noramullen1) [inline stdlib/experimental/http/requests/requests.flux:180]:
```suggestion
// - headers: Set of key values pairs to include on the request.
```

File: stdlib/experimental/http/requests/requests.flux [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+// - body: body is the data to send with the request
```
Comment (noramullen1) [inline stdlib/experimental/http/requests/requests.flux:181]:
```suggestion
// - body: Data to send with the request.
```

File: stdlib/experimental/http/requests/requests.flux [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+// - config: config is the set of options to control how the request should be performed.
```
Comment (noramullen1) [inline stdlib/experimental/http/requests/requests.flux:182]:
```suggestion
// - config: Set of options to control how the request should be performed.
```

**Review by noramullen1 — COMMENTED** [review]
Minor help text cleanup

**Comment by nathanielc** [comment]
@noramullen1 Thanks updated!

@onelson Thanks for the thoughts.

>  redirect policy (follow or not)

This seems like something we can add easily when it comes up. My hope was to design a config structure that makes adding new configuration easy while not having to anticipate all use cases from the start.

> loading a custom CA (which is more desirable than skipping cert verification)

This one is also one we could add later but its a little trickier. In our cloud offering where would you load the CA keys? You could possibly store them in our secrets store but that seems a bit heavy handed. For OSS allowing the config to point to files on disk would work well. Again I expect we could add this at a later point.



> functions to help build urls (specifically to help URI encode path segments and query params)

The current `requests.do` will encode the query params for you so its not strictly needed with the current API.

>  something to base64 encode strings (commonly needed to prepare headers/bodies)

I like this idea and agree a separate package is a likely a good place to do this. 


> I think there is a separate ticket out there related to hashing (md5, sha) which is often needed for making REST API requests. Common need is to hash some parts of the request params with the current timestamp, so it has to be done as close to the request being sent as possible.

I am not sure how we would do this one well. We have the ability to get the current time of a running Flux script, however it has some unique behaviors to Flux that might make this hard. Flux likes to pretend that the entire script runs in the same instant. This way if you are query data for different ranges you don't get slightly differing time ranges because of when the query plan itself was constructed. Not sure how I would handle this case.

**Comment by nathanielc** [comment]
> Commonly I've seen params expressed as a sequence of tuple pairs. If a map of string: [string] makes more sense for flux, then I'd take it, but in other langs I'd expect to be able to push new pairs into an indexed collection.

Flux doesn't have unnamed tuples, so you would either have to create a record for this purpose, like `[{key: string, value: string}]` or use a dictionary `[string:[string]]` As a result I think we should use the dictionary as it result in less verbose code and maintains good clarity for what is expected.

**Comment by onelson** [comment]
I totally agree that my wishlist of batteries to include can come later.

Small nitpick on

> The current `requests.do` will encode the query params for you so its not strictly needed with the current API.

One common place not covered by this behavior is in path segments. If they are using string interpolation to build their URL, they may need a way to encode the value... but again, we don't have to address all of this now. It's just the stuff I'd expect to find as a user.

**Comment by nathanielc** [comment]
I just finished the update to make `params` handle multiple values with the same key. I am realizing headers have the same behavior. Should I also change headers to be `[string:[string]]`? This one makes working with headers harder as they are both and input and output of these functions. Thoughts?

**Comment by onelson** [comment]
News to me if repeated header keys are valid -- the only time I've seen them in the wild, it was due to a configuration error on a reverse proxy, and it caused clients to have issues. I thought headers were not expected to be repeated. Client APIs normally encode headers as a flat map, but now I'm not sure what to think...

**Comment by onelson** [comment]
TIL https://stackoverflow.com/a/4371395

I'd only ever seen cache-control using the comma separated list form before today. Huh.

**Comment by nathanielc** [comment]
Thanks for the link, given that there is an equivalent way to construct the headers as CSV values then I think we should keep our API simpler and just support `[string:string]`. Then if a user needs multiple values they will be forced to use CSV formatting of the value which is acceptable. Also since order is important using CSV values make that more clear.

**Comment by nathanielc** [comment]
@onelson Thanks for all the feedback, can you give this final state a review before merging?

**Review by onelson — APPROVED** [review]
Looks good to me. The test cases hit the major notes and the docs read well.

**Pull request merged**